crates/storage-query-datafusion/src/remote_query_scanner_manager.rs

```diff
@@ -106,15 +129,42 @@ impl RemoteScannerManager {
     pub fn get_partition_target_node(
         &self,
         table: &str,
-        _partition_id: PartitionId,
-    ) -> PartitionLocation {
-        // TODO: obtain this information from the metadata, once exposed
-        PartitionLocation::Local {
-            scanner: self
-                .local_partition_scanner(table)
-                .expect("should be present"),
+        partition_id: PartitionId,
+    ) -> anyhow::Result<PartitionLocation> {
+        let my_node_id = self.my_node_id();
+
+        match self.partition_routing.get_node_by_partition(partition_id) {
+            None => {
+                bail!("node lookup for partition {} failed", partition_id)
+            }
+            Some(node_id)
+                if node_id
+                    .as_generational()
+                    .is_some_and(|id| id == *my_node_id) =>
+            {
+                Ok(PartitionLocation::Local {
+                    scanner: self
+                        .local_partition_scanner(table)
+                        .expect("should be present"),
+                })
+            }
+            Some(node_id) => Ok(PartitionLocation::Remote { node_id }),
         }
     }
+
+    // Note: we initialize my_node_id lazily as it is not available at construction time. We should
+    // be careful since the call to metadata().my_node_id() will panic if the system configuration
+    // isn't fully initialized yet, but we shouldn't be able to accept queries until that is in fact
+    // the case.
+    #[inline]
+    fn my_node_id(&self) -> &GenerationalNodeId {
+        self.my_node_id.get_or_init(|| {
+            self.metadata
+                .as_ref()
+                .expect("must have metadata if no node id is set")
+                .my_node_id()
+        })
```

Comment:
Why not always using `metadata.my_node_id()`?

Reply:
There is no functional need to cache it here, but @igalshilman is already taking care of this under his PR.